upstairs/src/volume.rs

```diff
@@ -3485,60 +3506,221 @@ mod test {
         }
     }
 
+    // For tests to decide what mix of read_only_parents the two VCRs should
+    // be created with.
+    #[derive(Debug, PartialEq)]
+    enum ReadOnlyParentMode {
+        Both,
+        OnlyOriginal,
+        OnlyReplacement,
+        Neither,
+    }
+
+    impl ReadOnlyParentMode {
+        fn test_all<F: FnMut(Self)>(mut f: F) {
+            f(ReadOnlyParentMode::Both);
+            f(ReadOnlyParentMode::OnlyOriginal);
+            f(ReadOnlyParentMode::OnlyReplacement);
+            f(ReadOnlyParentMode::Neither);
+        }
+    }
+
     #[test]
     fn volume_replace_basic() {
         // A valid replacement VCR is provided with only one target being
         // different.
         // Test all three targets for replacement.
-        for cid in 0..3 {
-            test_volume_replace_cid(cid).unwrap();
+        // Test with 1, 2, and 3 sub_volumes.
+        // Test with the difference being in each sub_volume.
+        // Test all combinations of read_only_parents.
+        for sv in 1..4 {
+            for sv_changed in 0..sv {
+                for cid in 0..3 {
+                    test_volume_replace_inner(
+                        sv,
+                        sv_changed,
+                        cid,
+                        ReadOnlyParentMode::Both,
+                    )
+                    .unwrap();
+                    test_volume_replace_inner(
+                        sv,
+                        sv_changed,
+                        cid,
+                        ReadOnlyParentMode::OnlyOriginal,
+                    )
+                    .unwrap();
+                    test_volume_replace_inner(
+                        sv,
+                        sv_changed,
+                        cid,
+                        ReadOnlyParentMode::OnlyReplacement,
+                    )
+                    .unwrap_err();
+                    test_volume_replace_inner(
+                        sv,
+                        sv_changed,
+                        cid,
+                        ReadOnlyParentMode::Neither,
+                    )
+                    .unwrap();
+                }
+            }
         }
     }
 
-    fn test_volume_replace_cid(cid: usize) -> Result<()> {
+    // Helper function to build as many sv_count sub-volumes as requested.
+    fn build_subvolume_vcr(
+        sv_count: usize,
+        block_size: u64,
+        blocks_per_extent: u64,
+        extent_count: u32,
+        opts: CrucibleOpts,
+        gen: u64,
+    ) -> Vec<VolumeConstructionRequest> {
+        (0..sv_count)
+            .map(|_| VolumeConstructionRequest::Region {
+                block_size,
+                blocks_per_extent,
+                extent_count,
+                opts: opts.clone(),
```

Comment:
The targets should be randomized here so that all sub volumes are not the same, using something like https://docs.rs/ipgen/1.0.2/ipgen/fn.ip.html

Reply:
I'm holding it wrong, somehow...

I get back:
`Err(InvalidIpNetwork("[BUG] failed to parse the generated IP address `a00:67` as IPv4"))`

And, the source says:
```
    // This error should never happen but I'm not a fan of panicking in libraries
    .map_err(|_| Error::InvalidIpNetwork(format!("[BUG] failed to parse the generated IP address `{}` as IPv4", ipv6_addr.trim_start_matches(':'))))
```

Reply:
Ah, it is a bug in the latest release 1.0.2. 
Fixed in main, but they have not rolled a new release out that contains the fix.